pkg/config/store/fsstore2.go

```diff
@@ -0,0 +1,228 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+)
+
+const defaultDuration = time.Second / 2
+
+type resourceMeta struct {
+	Name      string
+	Namespace string
+}
+
+type resource struct {
+	Kind       string
+	APIVersion string `json:"apiVersion"`
+	Metadata   resourceMeta
+	Spec       map[string]interface{}
+	sha        [sha1.Size]byte
+}
+
+func (r *resource) Key() Key {
+	return Key{Kind: r.Kind, Namespace: r.Metadata.Namespace, Name: r.Metadata.Name}
+}
+
+// fsStore2 is Store2Backend implementation using filesystem.
+type fsStore2 struct {
+	root          string
+	kinds         map[string]bool
+	checkDuration time.Duration
+	watchCtx      context.Context
+	watchCh       chan BackendEvent
+
+	mu   sync.RWMutex
+	data map[Key]*resource
+}
+
+var _ Store2Backend = &fsStore2{}
+
+func parseFile(path string, data []byte) []*resource {
+	if bytes.HasPrefix(data, []byte("---\n")) {
+		data = data[4:]
+	}
+	if bytes.HasSuffix(data, []byte("\n")) {
+		data = data[:len(data)-1]
+	}
+	if bytes.HasSuffix(data, []byte("\n---")) {
+		data = data[:len(data)-4]
+	}
+	if len(data) == 0 {
+		return nil
+	}
+	chunks := bytes.Split(data, []byte("\n---\n"))
+	resources := make([]*resource, 0, len(chunks))
+	for i, chunk := range chunks {
+		r := &resource{}
+		if err := yaml.Unmarshal(chunk, r); err != nil {
+			glog.Errorf("Failed to parse %d-th part in file %s: %v", i, path, err)
+			continue
+		}
+		if r.Kind == "" || r.Metadata.Namespace == "" || r.Metadata.Name == "" {
+			glog.Errorf("Key elements are empty. Extracted as %s", r.Key())
+			continue
+		}
+		r.sha = sha1.Sum(chunk)
+		resources = append(resources, r)
+	}
+	return resources
+}
+
+func (s *fsStore2) readFiles() map[Key]*resource {
+	const suffix = ".yaml"
+	result := map[Key]*resource{}
+
+	err := filepath.Walk(s.root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !strings.HasSuffix(path, suffix) || (info.Mode()&os.ModeType) != 0 {
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.Warningf("Failed to read %s: %v", path, err)
+			return err
+		}
+		for _, r := range parseFile(path, data) {
+			k := r.Key()
+			if !s.kinds[k.Kind] {
+				glog.Warningf("Unknown kind %s is in the file", k.Kind)
+				continue
+			}
+			result[r.Key()] = r
+		}
+		return nil
+	})
+	if err != nil {
+		glog.Errorf("failure during filepath.Walk: %v", err)
+	}
+	return result
+}
+
+func (s *fsStore2) update() {
```

Comment:
Rename:  `checkForUpdates()`

Reply:
This does update the content too. `checkAndUpdate` might be better?